projects/packages/stats/src/class-main.php

```diff
@@ -0,0 +1,217 @@
+	 * @param mixed $user_id User ID.
+	 * @return array Possibly mapped capabilities for meta capability.
+	 */
+	public static function map_meta_caps( $caps, $cap, $user_id ) {
+		// Map view_stats to exists.
+		if ( 'view_stats' === $cap ) {
+			$user        = new WP_User( $user_id );
+			$user_role   = array_shift( $user->roles );
+			$stats_roles = Options::get_option( 'roles' );
+
+			// Is the users role in the available stats roles?
+			if ( is_array( $stats_roles ) && in_array( $user_role, $stats_roles, true ) ) {
+				$caps = array( 'read' );
+			}
+		}
+
+		return $caps;
+	}
+
+	/**
+	 * Stats Template Redirect.
+	 *
+	 * @access public
+	 * @return void
+	 */
+	public static function template_redirect() {
+		global $current_user;
+
+		if (
+			is_feed()
+			|| is_robots()
+			|| is_embed()
+			|| is_trackback()
+			|| is_preview()
+			|| self::jetpack_is_dnt_enabled()
+		) {
+			return;
+		}
+
+		// @todo Is it ok to assume tracking will be enabled for any connected site?
+		if ( ! ( new Connection_Manager() )->is_connected() ) {
+			return;
+		}
+
+		// Staging Sites should not generate tracking stats.
+		$status = new Status();
+		if ( $status->is_staging_site() ) {
+			return;
+		}
+
+		// Should we be counting this user's views?
+		if ( ! empty( $current_user->ID ) ) {
+			$count_roles = Options::get_option( 'count_roles' );
+			if ( ! is_array( $count_roles ) || ! array_intersect( $current_user->roles, $count_roles ) ) {
+				return;
+			}
+		}
+
+		/**
+		 * Allow excluding specific IP addresses from being tracked in Stats.
+		 * Note: for this to work well, visitors' IP addresses must:
+		 * - be stored and returned properly in IP address headers;
+		 * - not be impacted by any caching setup on your site.
+		 *
+		 * @module stats
+		 *
+		 * @since-jetpack 10.6
+		 *
+		 * @param array $excluded_ips An array of IP address strings to exclude from tracking.
+		 */
+		$excluded_ips = (array) apply_filters( 'jetpack_stats_excluded_ips', array() );
+
+		// Should we be counting views for this IP address?
+		$current_user_ip = ( new Visitor() )->get_ip( true );
+		if (
+			! empty( $excluded_ips )
+			&& in_array( $current_user_ip, $excluded_ips, true )
+		) {
+			return;
+		}
+
+		add_action( 'wp_footer', array( 'Automattic\Jetpack\Stats\Tracking_Pixel', 'add_to_footer' ), 101 );
+		add_action( 'web_stories_print_analytics', array( 'Automattic\Jetpack\Stats\Tracking_Pixel', 'add_to_footer' ), 101 );
+	}
+
+	/**
+	 * CSS to hide the tracking pixel smiley.
+	 * It is now hidden for everyone (used to be visible if you had set the hide_smile option).
+	 *
+	 * @access public
+	 * @return void
+	 */
+	public static function hide_smile_css() {
+		?>
+	<style>img#wpstats{display:none}</style>
+		<?php
+	}
+
+	/**
+	 * Modified from stripslashes_deep()
+	 *
+	 * @todo Find a better home for this. Do we really need it though?
+	 *
+	 * @param mixed $value Value.
+	 * @return string
+	 */
+	public static function esc_html_deep( $value ) {
```

Comment:
Minor: Not crazy about this method, can't think of a proper home and I prefer to avoid Helper classes as they tend to get overloaded with irrelevant stuff that don't fit anywhere else.